src/models/chan.js

```diff
@@ -136,11 +136,27 @@ Chan.prototype.removeUser = function(user) {
 	this.users.delete(user.nick.toLowerCase());
 };
 
-Chan.prototype.toJSON = function() {
-	var clone = _.clone(this);
-	clone.users = []; // Do not send user list, the client will explicitly request it when needed
-	clone.messages = clone.messages.slice(-100);
-	return clone;
+Chan.prototype.getFilteredClone = function(lastActiveChannel, lastMessage) {
+	// Perform manual cloning of channel object for better control of performance and memory usage
+	return Object.keys(this).reduce((newChannel, prop) => {
+		if (prop === "users") {
+			// Do not send users, client requests updated user list whenever needed
+			newChannel[prop] = [];
+		} else if (prop === "messages") {
+			// If client is reconnecting, only send new message that client has not seen yet
+			if (lastMessage > -1) {
+				newChannel[prop] = this[prop].filter((m) => m.id > lastMessage);
+			}
+
+			// If channel is active, send up to 100 last messages, for all others send just 1
+			// Client will automatically load more messages whenever needed based on last seen messages
+			newChannel[prop] = this[prop].slice(this.id === lastActiveChannel ? -100 : -1);
+		} else {
+			newChannel[prop] = this[prop];
+		}
+
+		return newChannel;
+	}, {});
```

Comment:
Could you add tests for all those please? I'm really uncomfortable shipping convoluted code like this (convoluted for a reason, I know) without having it tested. Also, they're both part of the models, so that's not too tricky to test (i.e. there is no `irc` or `socket` object involved...).

Reply:
Add tests.